Parallel Hcurl NURBS space and an example named nurbs_25p

kaipengliuHIT opened the pull request:
After some minor changes, parallel Hcurl NURBS space IGA supported. And add an example nurbs_25p, it is the parallel nurbs version of the example 25. the resualt can be shown by paraview:
![paraview](https://github.com/user-attachments/assets/ed5ea0ca-655b-4bab-acfa-cc6f399b834b)

<!--GHEX{"author":"kaipengliuHIT","assignment":"2024-08-14T02:12:47","editor":"tzanio","id":4441,"reviewers":["justinlaughlin","dylan-copeland"],"merge":"2024-09-04T02:12:47","approval":"2024-08-28T02:12:47"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4441](https://github.com/mfem/mfem/pull/4441) | @kaipengliuHIT | @tzanio | @justinlaughlin + @dylan-copeland | 8/13/24 | ⌛due 8/27/24 | ⌛due 9/3/24 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

tzanio commented:
@kaipengliuHIT, let us know when this is `ready-for-review`

kaipengliuHIT commented:
> @kaipengliuHIT, let us know when this is `ready-for-review`

Yes, this is "ready-for-review". But I found that I can not add a label,   could you help me ? Thank you :)

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

dylan-copeland commented on fem/pfespace.cpp:
```diff
@@ -105,7 +105,15 @@ void ParFiniteElementSpace::ParInit(ParMesh *pm)
+      MFEM_ASSERT(own_ext, "internal error");
+      ParNURBSExtension *pNe = new ParNURBSExtension(NURBSext, VNURBSext,
+                                                     dynamic_cast<ParNURBSExtension *>(pmesh->NURBSext));
+      NURBSext = pNe;
```
Comment that `NURBSext` is deleted by the `ParNURBSExtension` constructor?

kaipengliuHIT replied:
OK

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -5000,6 +5000,181 @@ ParNURBSExtension::ParNURBSExtension(NURBSExtension *parent,
+   Swap(weights, parent->weights);
+   MFEM_VERIFY(!parent->HavePatches(), "");
+
+   delete parent;//delete series (may be series)
```
What does `series` mean here?

kaipengliuHIT replied:
the parent is constructed by series

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -5000,6 +5000,181 @@ ParNURBSExtension::ParNURBSExtension(NURBSExtension *parent,
+   int offset1 = 0;
+   int offset2 = 0;
+   int gndofs = 0;
+   if (VNURBSExt.Size() == 2)
```
If the size is neither 2 nor 3, should there be an error?

kaipengliuHIT replied:
I added a condition check.

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -5000,6 +5000,181 @@ ParNURBSExtension::ParNURBSExtension(NURBSExtension *parent,
+
+   // Only Support for the case when 'parent' is a local NURBSExtension.
+   // Construct the gtopo and ldof_group
+   MFEM_ASSERT(VNURBSExt.Size()!=0,"No VNURBSext !!! from pfespace.cpp row 5069");
```
Remove this? `from pfespace.cpp row 5069`

kaipengliuHIT replied:
OK

kaipengliuHIT replied:
I added a condition check.

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -5000,6 +5000,181 @@ ParNURBSExtension::ParNURBSExtension(NURBSExtension *parent,
+   }
+
+   Table dof_proc;
+   mfem::Array<int> partition = par_parent->partitioning;
```
`mfem::Array` -> `Array`

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -932,6 +938,12 @@ class ParNURBSExtension : public NURBSExtension
+                     Array<NURBSExtension *> VNURBSExt,
+                     const ParNURBSExtension *par_parent);
+
+   Array<int> GetPartitioning() {return partitioning;}
```
This is never called, so it can be removed (unless there is a good reason to keep it).

kaipengliuHIT replied:
removed

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -5377,4 +5552,144 @@ void NURBSPatchMap::SetBdrPatchDofMap(int p, const KnotVector *kv[],  int *okv)
    }
 }
 
+Table *NURBSExtension::GetGlobalElementDofTable()
```
These functions (1, 2, and 3D) all appear to be copied exactly from `ParNURBSExtension`. Would it be simpler to move them to `NURBSExtension`, so that `ParNURBSExtension` inherits them from the base class, instead of having identical copies in the base and derived classes?

kaipengliuHIT replied:
I found that due to the difference in the arrangement of degrees of freedom, this inheritance might become more complicated.

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,753 @@
+   VectorFunctionCoefficient E_Re(dim, E_bdr_data_Re);
+   VectorFunctionCoefficient E_Im(dim, E_bdr_data_Im);
+
+   mfem::out<<"main.cpp: test row 282"<<std::endl;
```
Remove

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,753 @@
+
+   ParFiniteElementSpace *fespace = new ParFiniteElementSpace(pmesh, NURBSext,
+                                                              fec);
+   cout << "Number of finite element unknowns: "
```
Begin the output with the MPI rank `myid`? Also in the `cout` calls below.

kaipengliuHIT replied:
add the myid

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,753 @@
+   prec.FormSystemMatrix(ess_tdof_list, PCOpAh);
+
+
+   // Gauss-Seidel Smoother
```
Is Gauss-Seidel an accurate description for this block diagonal AMG preconditioner?

kaipengliuHIT replied:
Sorry, I will correct it now.

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,753 @@
+   x.imag().Save(sol_i_ofs);
+
+   ParaViewDataCollection *pd = NULL;
+   pd = new ParaViewDataCollection("nurbs_ex25p", pmesh);
```
Instead of `new/delete`, this can simply be an object: `ParaViewDataCollection pd(...`

kaipengliuHIT replied:
Yes，however, I noticed that many other cases also use new.

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,753 @@
+   // NURBSext have been destoryed when construct the ParNURBSext!!!
+   if (myid == 0)
+   {
+      mfem::out<<"****** FINISH ******"<<std::endl;
```
Yes

kaipengliuHIT commented on mesh/nurbs.hpp:
```diff
@@ -932,6 +938,12 @@ class ParNURBSExtension : public NURBSExtension
    ParNURBSExtension(NURBSExtension *parent,
                      const ParNURBSExtension *par_parent);
+
+   ParNURBSExtension(NURBSExtension *parent,
```
Yes， I add the Document

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,753 @@
+//                                MFEM Example nurbs_25p -- modified for NURBS FE
+//
+// Compile with: make nurbs_ex25p
```
add the nurbs_ex25p in makefile

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,753 @@
+
+   pmesh->Print(mesh_ofs);
+
+   ofstream sol_r_ofs("nurbs_ex25p-sol_r.gf");
```
I revised them

tzanio commented:
ping: @kaipengliuHIT

kaipengliuHIT commented:
The original request only involved modifications related to IGA parallelism in the H(curl) space, etc., but I later added upgrades such as examples of GPU-based computations, unidirectional light sources, and multi-frequency calculation methods for time-varying systems. However, the programs for these latter algorithms do not yet comply with the standards, so I will revise them when I have time. For now, I will first close this pull request and create a new one to submit the earlier modifications for the H(curl) space geometric analysis parallelism. @tzanio @dylan-copeland

kaipengliuHIT commented:
Hi MFEM team,

I have updated this PR with the following changes:

**Purpose**: Add parallel H(curl) NURBS support for isogeometric electromagnetic optimization.

**Key modifications**:
- Added `VNURBSext` support in `ParFiniteElementSpace` (fem/pfespace.cpp)
- Extended `ParNURBSExtension` constructor to handle vector NURBS spaces for H(curl)
- Added global element DOF table methods in `NURBSExtension` (mesh/nurbs.hpp, mesh/nurbs.cpp)
- Added `nurbs_ex25p` example demonstrating parallel H(curl) NURBS usage

**CI Status**:
- ✅ Code style checks pass
- ✅ Branch history is now clean (2 commits, 7 files)
- ⚠️ Documentation warnings in [mesh/ncnurbs.cpp](file:///home/kaipeng/code/mfem/mesh/ncnurbs.cpp) are pre-existing and not introduced by this PR

The branch is now based on the latest `master` and ready for review. Please let me know if you need any clarifications or changes.

Thank you!
